testutil/eventually.go

```diff
@@ -0,0 +1,64 @@
+package testutil
+
+import (
+	"context"
+	"testing"
+	"time"
+)
+
+// Eventually is like require.Eventually except it takes a context.
+// If ctx times out, the test will fail.
+// ctx must have a deadline set or this will panic.
+func Eventually(ctx context.Context, t testing.TB, condition func() bool, tick time.Duration) bool {
+	t.Helper()
+
+	if _, ok := ctx.Deadline(); !ok {
+		panic("developer error: must set deadline on ctx")
+	}
+
+	ch := make(chan bool, 1)
+	ticker := time.NewTicker(tick)
+	defer ticker.Stop()
+	for tick := ticker.C; ; {
+		select {
+		case <-ctx.Done():
+			t.Errorf("Await timed out")
+			return false
+		case <-tick:
+			tick = nil
+			go func() { ch <- condition() }()
```

Comment:
We could consider not doing this in a goroutine, so that using `require` in `condition` func is allowed. The downside is we can't exit immediately on timeout but we could more prominently promote that this is the responsibility of the developer by making `condition` take a `context` as well (most likely just the same context passed to `Eventually` unless we find other reasons for cancellation that could be controlled by `Eventually`).

Reply:
👍  I like the idea of passing in `ctx`.